### DISCUSSION TITLE
Sidebar UI Element

### DISCUSSION THREAD

**Discussion by KristofPusztai** [marimo-team/marimo#1094] [Ideas]
Any plans to add a sidebar UI element? Something similar to the sidebar that's already present on the examples found in the [public marimo examples](https://marimo.io/@public)!

**Comment by mscolnick** [comment]
@KristofPusztai , you might be able try the Grid Layout and create it yourself: https://docs.marimo.io/guides/apps.html#layout

If we did have a sidebar, how would you like to configure it in code?

**Reply by KristofPusztai** [reply]
How would I go about doing this in Grid Mode? I was thinking something similar to the way [streamlits sidebars](https://docs.streamlit.io/develop/api-reference/layout/st.sidebar) work, getting something like this in grid mode using the `style` method to change the element background and position to be fixed, etc… But I havent tried this yet.

Getting such an element to have more dynamic functionality such as collapsing etc, would be similar to the already existing accordion component, just horizontal expansion instead of vertical!

**Reply by mscolnick** [reply]
With grid mode, it might be harder to have all the features like collapse while making it feel natural.

We will likely add some APIs around layout + multi-page + routing. I am curious if you have thoughts on how you'd like to interact with this.

E.g. 

```python
mo.page({
 sidebar=mo.vstack([
    mo.navitem(route="/dashboard", label="Dashboard"),
    mo.navitem(route="/settings", label="Settings"),
 ]),
 body=mo.route({
    "/dashboard": ...,
    "/settings": ...,
 }),
})

```

**Reply by KristofPusztai** [reply]
Yea something like that could be interesting as well for multi-page navigation, I hadn't quite thought about the sidebar as a routing tool though although that would probably be a main benefit. I was thinking of it more for displaying controls in a fixed way, that way as you scroll down the page, the sidebar remains on the side with various controls. 

I think your example makes sense although I'm not sure I understand how the *sidebar* and *body* args for this `mo.page` function would work together. Keeping in line with the current marimo ui api, for controls I was thinking of maybe something like:

```python
# Defining controls/any other individual elements
control1 = mo.ui.number(...)
control2 = mo.ui.dropdown(...)

# Defining sidebar
mo.ui.sidebar(
    title='Sidebar Title',
    body=mo.md(f"Control 1: {control1}  Control 2: {control2}")
    collapsible=True
)
```
This way the body can be formatted flexibly and it's up to the user on how to format the body of the sidebar element.

Example in mind:
![image](https://github.com/marimo-team/marimo/assets/52759282/3738cb7b-9c27-441f-b1bb-7893d3f0dbf5)
